File: Example/FWFramework/Classes/Module/Test/Plugin/TestPluginController.m
//
//  TestPluginController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/25.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestPluginController.h"
#import "AppSwift.h"
@import FWFramework;
@import Lottie;

@interface TestPluginController () <FWTableViewController>

@end

@implementation TestPluginController

- (UITableViewStyle)setupTableStyle
{
    return UITableViewStyleGrouped;
}

- (void)setupTableView
{
    [self.tableData addObject:@[@0, @1]];
    [self.tableData addObject:@[@0, @1, @2, @3, @4, @5, @6]];
    [self.tableData addObject:@[@0]];
    [self.tableView reloadData];
}

#pragma mark - TableView

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return self.tableData.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    NSArray *sectionData = [self.tableData objectAtIndex:section];
    return sectionData.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSArray *sectionData = [self.tableData objectAtIndex:indexPath.section];
    NSInteger rowData = [sectionData[indexPath.row] fw_safeInteger];
    if (indexPath.section == 0) {
        UITableViewCell *cell = [UITableViewCell fw_cellWithTableView:tableView style:UITableViewCellStyleDefault reuseIdentifier:@"cell1"];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        FWProgressView *view = [cell viewWithTag:100];
        if (!view) {
            view = [[FWProgressView alloc] init];
            view.tag = 100;
            view.color = AppTheme.textColor;
            [cell.contentView addSubview:view];
            view.fw_layoutChain.center();
        }
        view.annular = rowData == 0 ? YES : NO;
        [TestController mockProgress:^(double progress, BOOL finished) {
            view.progress = progress;
        }];
        return cell;
    }
    
    if (indexPath.section == 2) {
        UITableViewCell *cell = [UITableViewCell fw_cellWithTableView:tableView style:UITableViewCellStyleDefault reuseIdentifier:@"cell3"];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        FWLottieView *view = [cell viewWithTag:100];
        if (!view) {
            view = [[FWLottieView alloc] init];
            [view setAnimationWithName:@"Lottie" bundle:nil];
            view.tag = 100;
            view.color = AppTheme.textColor;
            [cell.contentView addSubview:view];
            view.fw_layoutChain.center();
        }
        [view startAnimating];
        return cell;
    }
    
    UITableViewCell *cell = [UITableViewCell fw_cellWithTableView:tableView style:UITableViewCellStyleDefault reuseIdentifier:@"cell2"];
    cell.selectionStyle = UITableViewCellSelectionStyleNone;
    FWIndicatorView *view = [cell viewWithTag:100];
    if (!view) {
        view = [[FWIndicatorView alloc] init];
        view.tag = 100;
        view.color = AppTheme.textColor;
        [cell.contentView addSubview:view];
        view.fw_layoutChain.center();
    }
    view.type = (FWIndicatorViewAnimationType)rowData;
    [view startAnimating];
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 70;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    FWWeakifySelf();
    [self fw_showAlertWithTitle:@"请选择" message:nil style:FWAlertStyleDefault cancel:nil actions:@[@"预览", @"设置全局样式"] actionBlock:^(NSInteger index) {
        FWStrongifySelf();
        if (index == 0) {
            [self onPreview:indexPath];
        } else {
            [self onSettings:indexPath];
        }
    } cancelBlock:nil];
}

#pragma mark - Action

- (void)onPreview:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [self.tableView cellForRowAtIndexPath:indexPath];
    if (indexPath.section == 0) {
        FWProgressView *view = [cell viewWithTag:100];
        [TestController mockProgress:^(double progress, BOOL finished) {
            view.progress = progress;
        }];
        return;
    }
    
    if (indexPath.section == 2) {
        FWToastPluginImpl *toastPlugin = [[FWToastPluginImpl alloc] init];
        toastPlugin.customBlock = ^(FWToastView * _Nonnull toastView) {
            FWLottieView *view = [[FWLottieView alloc] init];
            [view setAnimationWithName:@"Lottie" bundle:nil];
            toastView.indicatorView = view;
        };
        self.tableView.hidden = YES;
        [toastPlugin showLoadingWithAttributedText:[[NSAttributedString alloc] initWithString:@"Loading..."] cancelBlock:nil inView:self.view];
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [toastPlugin showLoadingWithAttributedText:[[NSAttributedString alloc] initWithString:@"Authenticating..."] cancelBlock:nil inView:self.view];
        });
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(3 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [toastPlugin hideLoading:self.view];
            self.tableView.hidden = NO;
        });
        return;
    }
    
    NSArray *sectionData = [self.tableData objectAtIndex:indexPath.section];
    FWIndicatorViewAnimationType type = [sectionData[indexPath.row] fw_safeInteger];
    FWToastPluginImpl *toastPlugin = [[FWToastPluginImpl alloc] init];
    toastPlugin.customBlock = ^(FWToastView * _Nonnull toastView) {
        toastView.indicatorView = [[FWIndicatorView alloc] initWithType:type];
    };
    self.tableView.hidden = YES;
    [toastPlugin showLoadingWithAttributedText:[[NSAttributedString alloc] initWithString:@"Loading..."] cancelBlock:nil inView:self.view];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [toastPlugin showLoadingWithAttributedText:[[NSAttributedString alloc] initWithString:@"Authenticating..."] cancelBlock:nil inView:self.view];
    });
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(3 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [toastPlugin hideLoading:self.view];
        self.tableView.hidden = NO;
    });
}

- (void)onSettings:(NSIndexPath *)indexPath
{
    if (indexPath.section == 0) {
        BOOL annular = indexPath.row == 0;
        FWViewPluginImpl.sharedInstance.customProgressView = ^UIView<FWProgressViewPlugin> * _Nonnull(FWProgressViewStyle style) {
            FWProgressView *progressView = [[FWProgressView alloc] init];
            progressView.annular = annular;
            return progressView;
        };
        FWRefreshPluginImpl.sharedInstance.pullRefreshBlock = nil;
        FWRefreshPluginImpl.sharedInstance.infiniteScrollBlock = nil;
        return;
    }
    
    if (indexPath.section == 2) {
        FWViewPluginImpl.sharedInstance.customIndicatorView = ^UIView<FWIndicatorViewPlugin> * _Nonnull(FWIndicatorViewStyle style) {
            FWLottieView *lottieView = [[FWLottieView alloc] init];
            [lottieView setAnimationWithName:@"Lottie" bundle:nil];
            return lottieView;
        };
        // FWLottieView也支持进度显示
        FWViewPluginImpl.sharedInstance.customProgressView = ^UIView<FWProgressViewPlugin> * _Nonnull(FWProgressViewStyle style) {
            FWLottieView *lottieView = [[FWLottieView alloc] init];
            [lottieView setAnimationWithName:@"Lottie" bundle:nil];
            lottieView.hidesWhenStopped = NO;
            return lottieView;
        };
        // FWLottieView支持下拉进度显示
        FWRefreshPluginImpl.sharedInstance.pullRefreshBlock = ^(FWPullRefreshView * _Nonnull pullRefreshView) {
            FWLottieView *lottieView = [[FWLottieView alloc] initWithFrame:CGRectMake(0, 0, 54, 54)];
            lottieView.contentInset = UIEdgeInsetsMake(5, 5, 5, 5);
            [lottieView setAnimationWithName:@"Lottie" bundle:nil];
            lottieView.hidesWhenStopped = NO;
            [pullRefreshView setAnimationView:lottieView];
        };
        FWRefreshPluginImpl.sharedInstance.infiniteScrollBlock = ^(FWInfiniteScrollView * _Nonnull infiniteScrollView) {
            FWLottieView *lottieView = [[FWLottieView alloc] initWithFrame:CGRectMake(0, 0, 44, 44)];
            [lottieView setAnimationWithName:@"Lottie" bundle:nil];
            lottieView.hidesWhenStopped = NO;
            [infiniteScrollView setAnimationView:lottieView];
        };
        return;
    }
    
    NSArray *sectionData = [self.tableData objectAtIndex:indexPath.section];
    FWIndicatorViewAnimationType type = [sectionData[indexPath.row] fw_safeInteger];
    FWViewPluginImpl.sharedInstance.customIndicatorView = ^UIView<FWIndicatorViewPlugin> * _Nonnull(FWIndicatorViewStyle style) {
        return [[FWIndicatorView alloc] initWithType:type];
    };
    // FWIndicatorView也支持进度显示
    FWViewPluginImpl.sharedInstance.customProgressView = ^UIView<FWProgressViewPlugin> * _Nonnull(FWProgressViewStyle style) {
        FWIndicatorView *indicatorView = [[FWIndicatorView alloc] initWithType:type];
        indicatorView.hidesWhenStopped = NO;
        return indicatorView;
    };
    FWRefreshPluginImpl.sharedInstance.pullRefreshBlock = nil;
    FWRefreshPluginImpl.sharedInstance.infiniteScrollBlock = nil;
}

@end

File: Example/FWFramework/Classes/Module/Test/Plugin/TestPluginController.swift
//
//  TestPluginController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/9/16.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework

class TestPluginController: UIViewController, TableViewControllerProtocol {
    
    typealias TableElement = [Int]
    
    @StoredValue("showLottieProgress")
    static var showLottieProgress = true
    
    @StoredValue("viewPluginSection")
    static var viewPluginSection: Int = -1
    
    @StoredValue("viewPluginRow")
    static var viewPluginRow: Int = 0
    
    func setupNavbar() {
        app.setRightBarItem(UIBarButtonItem.SystemItem.action.rawValue, block: { [weak self] _ in
            self?.app.showSheet(title: nil, message: nil, actions: [!Self.showLottieProgress ? "开启Lottie进度展示" : "关闭Lottie进度展示", "修改指示器颜色为红色"], actionBlock: { index in
                if index == 0 {
                    Self.showLottieProgress = !Self.showLottieProgress
                } else {
                    IndicatorViewStyle.setIndicatorColor(.red, for: .all)
                    ProgressViewStyle.setIndicatorColor(.red, for: .all)
                }
            })
        })
    }
    
    func setupTableStyle() -> UITableView.Style {
        .grouped
    }
    
    func setupTableView() {
        tableData.append([0, 1])
        tableData.append([0, 1, 2, 3, 4, 5, 6])
        tableData.append([0])
        tableView.reloadData()
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return tableData.count
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let sectionData = tableData[section]
        return sectionData.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let sectionData = tableData[indexPath.section]
        let rowData = sectionData[indexPath.row]
        if indexPath.section == 0 {
            let cell = UITableViewCell.app.cell(tableView: tableView, style: .default, reuseIdentifier: "cell1")
            cell.selectionStyle = .none
            var view = cell.viewWithTag(100) as? ProgressView
            if view == nil {
                let progressView = ProgressView()
                view = progressView
                progressView.tag = 100
                progressView.indicatorColor = AppTheme.textColor
                cell.contentView.addSubview(progressView)
                progressView.app.layoutChain.center()
            }
            view?.annular = rowData == 0 ? true : false
            TestController.mockProgress { progress, finished in
                view?.progress = progress
            }
            return cell
        }
        
        if indexPath.section == 2 {
            let cell = UITableViewCell.app.cell(tableView: tableView, style: .default, reuseIdentifier: "cell3")
            cell.selectionStyle = .none
            var view = cell.viewWithTag(100) as? LottiePluginView
            if view == nil {
                let lottieView = LottiePluginView()
                view = lottieView
                lottieView.animateWhenProgress = !Self.showLottieProgress
                lottieView.tag = 100
                lottieView.setAnimation(name: "Lottie")
                lottieView.indicatorColor = AppTheme.textColor
                cell.contentView.addSubview(lottieView)
                lottieView.app.layoutChain.center()
            }
            view?.startAnimating()
            return cell
        }
        
        let cell = UITableViewCell.app.cell(tableView: tableView, style: .default, reuseIdentifier: "cell2")
        cell.selectionStyle = .none
        var view = cell.viewWithTag(100) as? IndicatorView
        if view == nil {
            let indicatorView = IndicatorView()
            view = indicatorView
            indicatorView.tag = 100
            indicatorView.indicatorColor = AppTheme.textColor
            cell.contentView.addSubview(indicatorView)
            indicatorView.app.layoutChain.center()
        }
        view?.type = IndicatorViewAnimationType(rawValue: rowData)
        view?.startAnimating()
        return cell
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        70
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: false)
        app.showAlert(title: "请选择", message: nil, style: .default, cancel: nil, actions: ["预览", "设置全局样式"]) { [weak self] index in
            if index == 0 {
                self?.onPreview(indexPath)
            } else {
                TestPluginController.viewPluginSection = indexPath.section
                TestPluginController.viewPluginRow = indexPath.row
                TestPluginController.onSettings(indexPath)
            }
        }
    }
    
    func onPreview(_ indexPath: IndexPath) {
        let cell = tableView.cellForRow(at: indexPath)
        if indexPath.section == 0 {
            let progressView = cell?.viewWithTag(100) as? ProgressView
            TestController.mockProgress { progress, finished in
                progressView?.progress = progress
            }
            return
        }
        
        if indexPath.section == 2 {
            let toastPlugin = ToastPluginImpl()
            toastPlugin.customBlock = { toastView in
                let lottieView = LottiePluginView()
                lottieView.animateWhenProgress = !Self.showLottieProgress
                lottieView.setAnimation(name: "Lottie")
                toastView.indicatorView = lottieView
            }
            tableView.isHidden = true
            toastPlugin.showLoading(attributedText: NSAttributedString(string: "Loading..."), attributedDetail: nil, cancelBlock: nil, customBlock: nil, in: self.view)
            DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) {
                toastPlugin.showLoading(attributedText: NSAttributedString(string: "Authenticating..."), attributedDetail: nil, cancelBlock: nil, customBlock: nil, in: self.view)
            }
            DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
                toastPlugin.hideLoading(delayed: false, in: self.view)
                self.tableView.isHidden = false
            }
            return
        }
        
        let sectionData = tableData[indexPath.section]
        let rowData = sectionData[indexPath.row]
        let type = IndicatorViewAnimationType(rawValue: rowData)
        let toastPlugin = ToastPluginImpl()
        toastPlugin.customBlock = { toastView in
            toastView.indicatorView = IndicatorView(type: type)
        }
        tableView.isHidden = true
        toastPlugin.showLoading(attributedText: NSAttributedString(string: "Loading..."), attributedDetail: nil, cancelBlock: nil, customBlock: nil, in: self.view)
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) {
            toastPlugin.showLoading(attributedText: NSAttributedString(string: "Authenticating..."), attributedDetail: nil, cancelBlock: nil, customBlock: nil, in: self.view)
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
            toastPlugin.hideLoading(delayed: false, in: self.view)
            self.tableView.isHidden = false
        }
    }
    
    static func onSettings(_ indexPath: IndexPath) {
        if indexPath.section == 0 {
            let annular = indexPath.row == 0
            ViewPluginImpl.shared.customProgressView = { style in
                let progressView = ProgressView()
                progressView.annular = annular
                return progressView
            }
            RefreshPluginImpl.shared.pullRefreshBlock = nil
            RefreshPluginImpl.shared.infiniteScrollBlock = nil
            return
        }
        
        if indexPath.section == 2 {
            ViewPluginImpl.shared.customIndicatorView = { style in
                let lottieView = LottiePluginView()
                lottieView.animateWhenProgress = !Self.showLottieProgress
                lottieView.setAnimation(name: "Lottie")
                return lottieView
            }
            // FWLottieView也支持进度显示
            ViewPluginImpl.shared.customProgressView = { style in
                let lottieView = LottiePluginView()
                lottieView.animateWhenProgress = !Self.showLottieProgress
                lottieView.setAnimation(name: "Lottie")
                lottieView.hidesWhenStopped = false
                return lottieView
            }
            // FWLottieView支持下拉进度显示
            RefreshPluginImpl.shared.pullRefreshBlock = { view in
                let lottieView = LottiePluginView(frame: CGRect(x: 0, y: 0, width: 54, height: 54))
                lottieView.animateWhenProgress = !Self.showLottieProgress
                lottieView.contentInset = UIEdgeInsets(top: 5, left: 5, bottom: 5, right: 5)
                lottieView.setAnimation(name: "Lottie")
                lottieView.hidesWhenStopped = false
                view.setAnimationView(lottieView)
            }
            RefreshPluginImpl.shared.infiniteScrollBlock = { view in
                let lottieView = LottiePluginView(frame: CGRect(x: 0, y: 0, width: 44, height: 44))
                lottieView.animateWhenProgress = !Self.showLottieProgress
                lottieView.setAnimation(name: "Lottie")
                lottieView.hidesWhenStopped = false
                view.setAnimationView(lottieView)
            }
            return
        }
        
        let type = IndicatorViewAnimationType(rawValue: indexPath.row)
        ViewPluginImpl.shared.customIndicatorView = { style in
            return IndicatorView(type: type)
        }
        // IndicatorView也支持进度显示
        ViewPluginImpl.shared.customProgressView = { style in
            let indicatorView = IndicatorView(type: type)
            indicatorView.hidesWhenStopped = false
            return indicatorView
        }
        RefreshPluginImpl.shared.pullRefreshBlock = nil
        RefreshPluginImpl.shared.infiniteScrollBlock = nil
    }
    
}

@objc extension Autoloader {
    func loadApp_ViewPlugin() {
        if TestPluginController.viewPluginSection >= 0 {
            TestPluginController.onSettings(IndexPath(row: TestPluginController.viewPluginRow, section: TestPluginController.viewPluginSection))
        }
    }
}
